### PR TITLE
unauthorize expired users

### PR DESCRIPTION
**Pull request by neexite**
[AB#68950](https://dev.azure.com/orckestra001/da965986-c85b-43be-9915-219568b000e5/_workitems/edit/68950)

### CODE REVIEW

**Comment by InnaBoitsun** [comment]
Can we somehow also clean frontend cache , so in header we don't My Account

image

File: src/Orckestra.Composer/Services/ComposerContext.cs [hunk 1]
```diff
@@ -243,9 +253,35 @@ private void InitializeCustomerId()
...
+                            Scope = Scope,
+                            CultureInfo = CultureInfo,
+                            IncludeAddresses =
+                                true // all parameters should be the same as in CustomerViewService.GetAccountHeaderViewModelAsync
+                        }).Result;
+                        
+                        var passwordChangedDateTime = customer.LastPasswordChanged.ToUniversalTime();
+                        // 10 seconds was added to compensate OCS server and EM server difference time.
+                        var ticketDateTime = (HttpContextBase.User?.Identity as System.Web.Security.FormsIdentity)?.Ticket
+                            .IssueDate.ToUniversalTime().AddSeconds(10);
```
Comment (napernik) [inline src/Orckestra.Composer/Services/ComposerContext.cs]: I suggest removing the 10 seconds, as it is potentially exploitable.

File: src/Orckestra.Composer/Services/ComposerContext.cs [hunk 2]
```diff
@@ -231,6 +237,10 @@ public virtual string GetEncryptedCustomerId()
             return dto.EncryptedCustomerId;
         }
 
+        // ReSharper disable once InconsistentNaming
+        private readonly bool ValidationPasswordChangedNotDisabled =
```
Comment (napernik) [inline src/Orckestra.Composer/Services/ComposerContext.cs]: You can call it something like **PasswordChangedValidationEnabled**, this way it is both shorter and ReSharper  won't complain

File: src/Orckestra.Composer/Services/ComposerContext.cs [hunk 3]
```diff
@@ -243,9 +253,35 @@ private void InitializeCustomerId()
...
+                    if (dto.IsGuest != true && ValidationPasswordChangedNotDisabled)
+                    {
+                        // GetCustomerByIdAsync uses a cache
+                        var customer = CustomerRepository.GetCustomerByIdAsync(new GetCustomerByIdParam()
+                        {
+                            CustomerId = decryptedCustomerId,
+                            Scope = Scope,
+                            CultureInfo = CultureInfo,
+                            IncludeAddresses =
+                                true // all parameters should be the same as in CustomerViewService.GetAccountHeaderViewModelAsync
```
Comment (napernik) [inline src/Orckestra.Composer/Services/ComposerContext.cs:266]: There should be no need to include addresses, as only LastPasswordChanged is used. Or do you need to hit the same cache entry?

Comment (neexite) [reply src/Orckestra.Composer/Services/ComposerContext.cs:266]: parameters the same  for cache entry,

Comment (napernik) [reply src/Orckestra.Composer/Services/ComposerContext.cs:266]: Oh.. the caching is not implemented correctly, there're multilple places where a customer is read with or without the IncludeAddress, but IncludeAddress is not a part of the cache key.
I suggest not to include address here, and then make another PR, where IncludeAddress is added as a part of cache key.

**Review by napernik — APPROVED** [review]

**Pull request merged**